core/store/src/flat_state.rs

```diff
@@ -726,6 +870,92 @@ mod tests {
+                flat_state_factory.new_flat_state_for_shard(0, Some(block_hash), false).unwrap();
+            assert_eq!(flat_state.get_ref(&[1]).unwrap(), Some(ValueRef::new(&[i as u8])));
+        }
+
+        // 3. Create a new block that deletes &[1] and add a new value &[2]
+        //    Add the block to flat storage.
+        let hash = chain.create_block();
+        let store_update = flat_storage_state
+            .add_block(
+                &hash,
+                FlatStateDelta::from([(vec![1], None), (vec![2], Some(ValueRef::new(&[1])))]),
+                chain.get_block_info(&hash),
+            )
+            .unwrap();
+        store_update.commit().unwrap();
+
+        // 4. Create a flat_state0 at block 10 and flat_state1 at block 4
+        //    Verify that they return the correct values
+        let deltas =
+            flat_storage_state.get_deltas_between_blocks(&chain.get_block_hash(10)).unwrap();
+        assert_eq!(deltas.len(), 11);
+        let flat_state0 = flat_state_factory
+            .new_flat_state_for_shard(0, Some(chain.get_block_hash(10)), false)
+            .unwrap();
+        let flat_state1 = flat_state_factory
+            .new_flat_state_for_shard(0, Some(chain.get_block_hash(4)), false)
+            .unwrap();
+        assert_eq!(flat_state0.get_ref(&[1]).unwrap(), None);
+        assert_eq!(flat_state0.get_ref(&[2]).unwrap(), Some(ValueRef::new(&[1])));
+        assert_eq!(flat_state1.get_ref(&[1]).unwrap(), Some(ValueRef::new(&[4])));
+        assert_eq!(flat_state1.get_ref(&[2]).unwrap(), None);
+
+        // 5. Move the flat head to block 5, verify that flat_state0 still returns the same values
+        // and flat_state1 returns an error. Also check that DBCol::FlatState is updated correctly
+        flat_storage_state.update_flat_head(&chain.get_block_hash(5)).unwrap();
+        assert_eq!(store_helper::get_state(&store, &[1]).unwrap(), Some(ValueRef::new(&[5])));
+        let deltas =
+            flat_storage_state.get_deltas_between_blocks(&chain.get_block_hash(10)).unwrap();
+        assert_eq!(deltas.len(), 6);
+        assert_eq!(flat_state0.get_ref(&[1]).unwrap(), None);
+        assert_eq!(flat_state0.get_ref(&[2]).unwrap(), Some(ValueRef::new(&[1])));
+        assert_matches!(flat_state1.get_ref(&[1]), Err(StorageError::FlatStorageError(_)));
+
+        // 6. Move the flat head to block 10, verify that flat_state0 still returns the same values
+        //    Also checks that DBCol::FlatState is updated correctly.
+        flat_storage_state.update_flat_head(&chain.get_block_hash(10)).unwrap();
+        assert_eq!(store_helper::get_state(&store, &[1]).unwrap(), None);
+        assert_eq!(store_helper::get_state(&store, &[2]).unwrap(), Some(ValueRef::new(&[1])));
+        assert_eq!(flat_state0.get_ref(&[1]).unwrap(), None);
+        assert_eq!(flat_state0.get_ref(&[2]).unwrap(), Some(ValueRef::new(&[1])));
+    }
+
     #[test]
     fn flat_state_apply_single_delta() {
```

Comment:
We can remove stubs `flat_state_apply_single_delta` and `flat_state_apply_delta_range`, because `flat_storage_state_sanity` cover all cases